### DISCUSSION TITLE
recognition training substring not found

### DISCUSSION THREAD

**Discussion by Atomme1** [mindee/doctr#926] [Q&A] [answered]
Hello, i found your repo last week and let me tell you that's one of the best OCR i've ever used ^^.
I have understand we can train a text recognition model and I wanted to try.

I have already looked in every discussions inside the repo so i'm adding new problems to the list yay :)

So i managed to charge my datasets test_path and val_path,  you can see it on the picture (don't mind the name of these files ), i used the pretrained option (and thus i had to change the vocab to legacy_french because the sizes didn't match but now it's fixed).

![pb_doctr_trainer](https://user-images.githubusercontent.com/43566264/169707701-cfcc2cf2-dcb3-4677-9158-c05b74128ef6.PNG)

There is something i don't get and i'm seeking for your help <3
Thank you in advance !!

NB: In the https://github.com/mindee/doctr/tree/main/references/recognition there is a little flaw on the «JSON file»
Because it needs to be like that 
-> 
{
    'img_1.jpg': 'I',
    'img_2.jpg': 'am',
    'img_3.jpg': 'a',
    'img_4.jpg': 'Jedi',
    'img_5.jpg': '!',
    ...
}
and not with the «labels=» 

I had to seek that one on the [#637](https://github.com/mindee/doctr/discussions/637) done by @fg-mindee

**Comment by frgfm** [comment]
Hi @Atomme1 :) 

Thanks a lot for sharing the appreciation, we're always glad to discuss with users 🙌 

About your problem, I'm not sure I get your problem. In #637 and the README of reference, we mention the same dict structure of `labels.json`. So what is the issue you are facing now? That despite following this, you encountered the error in your screenshot? 🤔 

Also, for next problems you encountered, here are two suggestions :
- if you encounter an error, that's a bug report (meaning it's better to open a GH issue https://github.com/mindee/doctr/issues/new?assignees=&labels=type%3A+bug&template=bug_report.yml
- every single time to have to share code & console output, try to copy caste the text :) rather than sharing a screenshot (to stay brief, if I have to reproduce what you've done, I prefer to copy paste rather than writing what I see in your screenshot. And also a screenshot is a much heavier file than pure text 😁 )

Cheers ✌️

**Reply by Atomme1** [reply]
Hello again !
I thought an image would've been better but now i understand why it isn't xD (sorry again)
And yes i should've opened a GH issue (my bad)

I had a problem with the JSON file format but i resolved it.

So yeah on the readME, the JSON file should have an architecture like that : 
labels = {
    'img_1.jpg': 'I',
    'img_2.jpg': 'am',
    'img_3.jpg': 'a',
    'img_4.jpg': 'Jedi',
    'img_5.jpg': '!'
}

but while trying that it showed me an error so i went and looked different bug report to see if someone already reported it.
And yes someone had the same problem (https://github.com/mindee/doctr/discussions/637).

So i just want to know if you could change the readME of the JSON part here https://github.com/mindee/doctr/tree/main/references/recognition
from being:
labels = {'img_1.jpg': 'I', 'img_2.jpg': 'am',.. }

to this:
{'img_1.jpg': 'I', 'img_2.jpg': 'am', .. }

In short term, get rid of the « labels = »

Cheers ✌️

**Comment by felixdittrich92** [comment, marked as answer]
@frgfm @Atomme1 
The problem is easy to describe: 

1. Yes it's a bug which can only be triggered while using the training scripts 
2. If your dataset does not contain a "letter/char in the labels" for EACH letter/char in the used VOCAB or on the other hand your dataset contains letters which are NOT in the used VOCAB it will be also raise this error 

Ref:
https://github.com/mindee/doctr/blob/main/doctr/datasets/utils.py 

@frgfm we have to modiy the encode_sequence function if char not in Vocab raise an error otherwise if not all chars from vocab in labels fill the missing ones with an unused index wdyt ?
:)

**Reply by Atomme1** [reply]
Heyy !

Thank you very much for the quick answer, i'm going to check that one out !

Have a wonderful day ! 👋

**Comment by felixdittrich92** [comment]
#927
